### PR TITLE
`Integrated code lifecycle`: Fix an issue when parsing special characters

### PR DESCRIPTION
**Pull request by Strohgelaender**
#### General
- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
When the result.xml file contains invalid characters (e.g. when they are used in stdout which is provided as part of the result file) the parsing fails.

Test that the newly changed C/GCC template works correctly.

### Description
We now remove such invalid characters before parsing.

### Steps for Testing
code review, quite hard to replicate since you need an exercise where such invalid chars get printed. Reviewing the new test case should be enough. You can also locally outcomment the replaceAll line and see that the test now fails.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

## Summary by CodeRabbit

- **New Features**
  - Introduced a function to sanitize XML inputs, ensuring invalid characters are removed.

- **Bug Fixes**
  - Broadened exception handling for parsing test results to handle any error, improving robustness.

- **Tests**
  - Added tests to verify the handling of invalid XML characters.

### CODE REVIEW

**Comment by upsj** [comment]
Maybe it also makes sense to address the source, which is (maybe among other places) https://github.com/ls1intum/Artemis/blob/bf3df4117cbe0cc4450a64a4f102b573056ecdab/src/main/resources/templates/c/gcc/test/testUtils/junit/TestCase.py#L47 where the stdout/stderr gets passed unsanitized to ElementTree, which leads to a broken XML file due to https://github.com/python/cpython/issues/49416

Unfortunately, my Python-foo is not strong enough to adapt the suggested workaround to modern Python versions

**Comment by Strohgelaender** [comment]
Thanks for the suggestion @upsj, I added the changes.

**Review by BBesrour — DISMISSED** [review]
Tested on TS1
There is an issue with the template
image
But I think this was also present before.
Also code lgtm

**Comment by upsj** [comment]
@Strohgelaender seems like the second line of the XML sanitation function was lost when applying the patch?

File: src/main/resources/templates/c/gcc/test/testUtils/Utils.py [hunk 1]
```diff
@@ -155,6 +156,14 @@ def printProg(text: str, addToCache: bool = True):
...
         testerOutputCache.append(msg)
 
+def sanitizeXml(text: str) -> str:
+    """
+    Removes all characters that are not allowed in XML
+    """
+    # remove ANSI escape sequences first to avoid spurious characters
+    # https://stackoverflow.com/questions/14693701/how-can-i-remove-the-ansi-escape-sequences-from-a-string-in-python
+    text = re.sub(r'\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])', "", text)
+    return re.sub(r'\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])', "", text)
```
[the remark this answers is not in the document]
Comment (upsj) [reply src/main/resources/templates/c/gcc/test/testUtils/Utils.py:166]: Based on my patch, this should be
```suggestion
    text = re.sub(r'\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])', "", text)
    return re.sub("[\\x01-\\x08\\x0B-\\x0C\\x0E-\\x1F\\x7F-\\x84\\x86-\\x9F]", "", text)
```

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by krusche — APPROVED** [review]

**Pull request merged**